src/socket-proxy/socket-proxyd.c

```
@@ -297,11 +300,89 @@ static int connection_enable_event_sources(Connection *c) {
+        if (getpeername(c->server_fd, (struct sockaddr*) &remote_addr, &addr_len) < 0) {
+                log_warning_errno(errno, "Failed to get remote address (getpeername)");
+                goto unknown;
+        }
+
+        switch (remote_addr.sa.sa_family) {
+        case AF_INET:
+                s_len = asprintf(&s, "PROXY TCP4 %s %s %hu %hu\r\n",
+                        IN4_ADDR_TO_STRING(&(remote_addr.in.sin_addr)),
+                        IN4_ADDR_TO_STRING(&(local_addr.in.sin_addr)),
+                        be16toh(remote_addr.in.sin_port),
+                        be16toh(local_addr.in.sin_port)
+                );
+                break;
+        case AF_INET6:
+                s_len = asprintf(&s, "PROXY TCP6 %s %s %hu %hu\r\n",
+                        IN6_ADDR_TO_STRING(&(remote_addr.in6.sin6_addr)),
+                        IN6_ADDR_TO_STRING(&(local_addr.in6.sin6_addr)),
+                        be16toh(remote_addr.in6.sin6_port),
+                        be16toh(local_addr.in6.sin6_port)
+                );
+                break;
+        default:
+                goto unknown;
+        }
+
+        if (s_len < 0)
+                goto unknown;
+        r = loop_write(c->client_fd, s, s_len);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to write to server: %m");
+                goto quit;
+        }
+
+        return;
+
+unknown:
+        /* ignore errors - server can decide to deny UNKNOWN connections */
+        r = loop_write(c->client_fd, "PROXY UNKNOWN\r\n", SIZE_MAX);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to write to server: %m");
+                goto quit;
+        }
+        return;
+
+quit:
+        /* client connection (to server) is unusable, close it */
+        connection_release(c);
+        return;
+}
+
 static int connection_complete(Connection *c) {
         int r;
 
         assert(c);
 
+        if (arg_proxy_protocol == 1)
+                send_proxy_protocol_v1(c);
```

Then, please propagate the error. The caller will call `connection_release()`.

---

ok I need to think about that (I'm only an amateur in C)